apps/prairielearn/assets/scripts/lib/countdown.ts

```diff
@@ -115,13 +124,22 @@ export function setupCountdown(options: {
     countdownDisplay.innerText = remainingSec >= 60 ? remainingMin + ' min' : remainingSec + ' sec';
 
     if (remainingMS > 0) {
+      if (
+        options.onRemainingTime &&
+        remainingTimeEvents.length > 0 &&
+        remainingMS < remainingTimeEvents[remainingTimeEvents.length - 1]
+      ) {
+        // need to provide fallback here although remainingTimeEvents.pop() will never fail
+        const event = remainingTimeEvents.pop() ?? 0;
```

Comment:
I'm not sure if this yields the desired behavior when the instructor manually extends time for an assessment instance. Consider the case where an instructor grants a student an additional 20 minutes when they only have 30 seconds to go. The student may have already dismissed this warning by this point. When the student gets back to 60 seconds left, this won't be shown again.

I won't block this PR on this, but let's at least discuss options here. Maybe start by sharing why you felt it was necessary to remove callbacks after they're called. Do you have any thoughts on how to handle this better?

Reply:
I haven't considered the case where the timer gets extended without a refresh of the page. The main reason I remove callbacks is to be able to use an inequality operator to check for them without having them triggered multiple times.

Note sure if there's a more elegant solution, but one alternative would be to flag them as processed, and add some logic for unflagging if a server sync returns a remaining time value that is greater than the previous sync's time.

Reply:
An option would be to instead trigger events based on the current remaining time and the remaining time at the last tick. If we start tracking the last tick, we can trigger the event when `lastRemaining < triggerRemaining <= currentRemaining` (I think, double-check my inequalities there).

Reply:
Changed to follow this suggestion. Thanks!